### PR TITLE
Add `CupyMatrices` and fix `asmatrix_fused`

### PR DESCRIPTION
**Pull request by andrea-pasquale**
This PR fixes qiboteam/qibo#928 by adding a class `CupyMatrices` and by also fixing the `asmatrix_fused` method.
This PR also addresses https://github.com/qiboteam/qibo/issues/928#issuecomment-1602391887 this second point raised by @renatomello.

It has been tested on GPU with qiboteam/qibo#950.

### CODE REVIEW

File: src/qibojit/backends/matrices.py
Comment (andrea-pasquale) [inline src/qibojit/backends/matrices.py:13]: I don't know if this is the best solution but it seems to work.
Perhaps we could improve readability by changing the name of the attribute `np`, even if I know that for specify methods we might need to access numpy also in backends different from the `NumpyBackend`. So I'm not sure.

**Review by stavros11 — APPROVED** [review]
Thanks, for implementing this. The fix for the fused gate looks good to me. 

For `CupyMatrices`, I saw the comments in the qibo issue but I am still not sure why we need them. I think the idea for having the gate matrices only in numpy was that they are small arrays (2x2 or 4x4) and it is typically faster to do these operations with numpy than casting to a GPU. We only cast them to cupy when we apply them to a state, like you do for the fused gate. I am not sure if this caused any bug though.

File: src/qibojit/backends/matrices.py
Comment (stavros11) [reply src/qibojit/backends/matrices.py:13]: To me it is fine, because I think if you change its name it will break `TensorflowMatrices`. But I'd still confirm that we really need the `CupyMatrices` before thinking about this because maybe we don't need to add this at all.

**Review by scarrazza — APPROVED** [review]
Thanks @andrea-pasquale, could you please update the qibo minimum required version to 0.1.15?

**Comment by andrea-pasquale** [comment]
@scarrazza I am guessing that you are planning to release qibo 0.1.15 after merging qiboteam/qibo#928, right?
Before that I want to check @stavros11 comment. I believe that there should be a way to make the code work removing `CupyMatrices`.

**Comment by scarrazza** [comment]
Yes, we should merge https://github.com/qiboteam/qibo/issues/928 first and then release qibojit.

**Comment by andrea-pasquale** [comment]
I try to rerun @renatomello's example

```py
from qibojit.backends import CupyBackend

from qibo import gates, matrices
from qibo.backends import NumpyBackend
from qibo.gates.special import FusedGate

nqubits = 1
I = matrices.I

for backend in [NumpyBackend, CupyBackend]:
    backend = backend()
    I = backend.cast(I, dtype=I.dtype)
    gate = FusedGate(*range(nqubits))
    gate.append(gates.Unitary(I, *range(nqubits)))
    gate = gate.asmatrix(backend)
    print(type(gate))
```
and I get the following error (once I remove `CupyMatrices`:

```sh
Traceback (most recent call last):
  File "/nfs/users/andrea.pasquale/qibo/test.py", line 15, in <module>
    gate = gate.asmatrix(backend)
  File "/nfs/users/andrea.pasquale/qibo/src/qibo/gates/special.py", line 99, in asmatrix
    return backend.asmatrix_fused(self)
  File "/nfs/users/andrea.pasquale/qibojit/src/qibojit/backends/gpu.py", line 187, in asmatrix_fused
    npmatrix = super().asmatrix_fused(gate)
  File "/nfs/users/andrea.pasquale/qibo/src/qibo/backends/numpy.py", line 122, in asmatrix_fused
    gmatrix = self.to_numpy(gate.asmatrix(self))
  File "/nfs/users/andrea.pasquale/qibo/src/qibo/gates/abstract.py", line 386, in asmatrix
    return backend.asmatrix_parametrized(self)
  File "/nfs/users/andrea.pasquale/qibo/src/qibo/backends/numpy.py", line 113, in asmatrix_parametrized
    return getattr(self.matrices, name)(*gate.parameters)
  File "/nfs/users/andrea.pasquale/qibo/src/qibo/backends/npmatrices.py", line 274, in Unitary
    return self.np.array(u, dtype=self.dtype, copy=False)
  File "cupy/_core/core.pyx", line 1480, in cupy._core.core._ndarray_base.__array__
TypeError: Implicit conversion to a NumPy array is not allowed. Please use `.get()` to c
onstruct a NumPy array explicitly.
```
I think the error is triggered by the fact if we consider a `Unitary` gate the gate's parameters contains the matrix itself, and we do not perform any check on the type of matrix that the user passes to `Unitary`. In fact I can trigger the same error with the following script:

```py
import cupy as cp
from qibo import gates, models
from qibo.backends import set_backend


set_backend("numpy")

gate = gates.Unitary(cp.array([[0,1],[1,2]]), 0)

circuit = models.Circuit(1)
circuit.add(gate)
circuit()

```
The initial issue can be fixed by adding `CupyMatrices` given that the `Unitary` matrix is defined in the following way
https://github.com/qiboteam/qibo/blob/232af9176c4e64882b66193f1aa4942e3dc3ed5a/src/qibo/backends/npmatrices.py#L273-L274
 which will convert the matrix to a cupy array given that I've redefined np to cp.

However, this solution doesn't fix the latest script that I wrote.

Another possible approach will be to check explicitly the type of matrix that the user is specifying when a `Unitary` gate
is created. A possible issue with this solution is that the `Unitary` gate will define backend dependent  (which is somehow expected given that one of the parameters is a matrix...)

Let me know which approach do you prefer @stavros11 @scarrazza

**Comment by stavros11** [comment]
I think the main reason behind these issues is that we don't have any rules for the types of gate matrices. One possibility is to use numpy arrays for all backends, since this is expected to be the fastest for small matrices. The issue with this is that if we do it for the tensorflow backend it will break the tensorflow graph and it won't be possible to backpropagate on the gates of variational circuits, which is probably the only reason one would use tensorflow in qibo. This is why we have the `TensorflowMatrices`.

For cupy/cuquantum we can either use numpy or cupy for gate matrices. If we use cupy I think your solution is fine. I think it is also possible to keep using numpy and fix the first example if in the `CupyMatrices` you only override the `Unitary` to something like
```py
def Unitary(self, u):
    if isinstance(u, cp.ndarray):
        u = u.get()
    return super().Unitary(u)
```
without doing `self.np = cp` (if you don't like this). Now `Unitary` and all other matrices will still be numpy.

For the second example, because qibo gates are backend agnostic, the moment you do `gates.Unitary(cp.array([[0,1],[1,2]]), 0)` you don't know if you are going to use this gate with cupy or a different backend so you cannot raise an error. That is unless we decide to have all gates in numpy (regardless of backend) and raise an error otherwise (but this won't work for tensorflow). But generally I wouldn't worry too much about this example, because if the user is advanced enough to initialize the unitary using a cupy array, they will probably know why it didn't work with the numpy backend. I would say the most common case is the opposite, initialize with numpy and use with cupy, which shouldn't throw any error.

**Comment by andrea-pasquale** [comment]
Thanks for the feedback @stavros11.

> One possibility is to use numpy arrays for all backends, since this is expected to be the fastest for small matrices.

If this is the case I would say that it is better to keep `numpy` also for cupy.
I will implement the solution that you proposed by overriding only the `Unitary` gate.




> For the second example, because qibo gates are backend agnostic, the moment you do `gates.Unitary(cp.array([[0,1],[1,2]]), 0)` you don't know if you are going to use this gate with cupy or a different backend so you cannot raise an error. That is unless we decide to have all gates in numpy (regardless of backend) and raise an error otherwise (but this won't work for tensorflow). But generally I wouldn't worry too much about this example, because if the user is advanced enough to initialize the unitary using a cupy array, they will probably know why it didn't work with the numpy backend. I would say the most common case is the opposite, initialize with numpy and use with cupy, which shouldn't throw any error.

I think that you are right. Maybe this particular example is a bit too complicated.

File: src/qibojit/backends/matrices.py
Comment (stavros11) [inline src/qibojit/backends/matrices.py:3]: I think if you put this import here, we will need cupy a requirement of qibojit, otherwise it will fail to import anything from qibojit, even if it is numba related only. Currently it is not a requirement so that we don't force users who don't have a GPU and only use qibojit-numba to install cupy.

Alternatively (probably better) we can just move the import inside `def Unitary:`

Comment (andrea-pasquale) [reply src/qibojit/backends/matrices.py]: You are right. I forgot that cupy is not a dep. This explains also why all the imports are inside classes/functions.

**Comment by scarrazza** [comment]
@andrea-pasquale thanks for the changes, could you please run tests at least once pointing again to https://github.com/qiboteam/qibo/pull/950?

**Comment by andrea-pasquale** [comment]
I've changed the qibo version from 0.1.15 to the branch `fix_fused`.
Tests on GPU are passing locally both for qibo on qibojit.

**Comment by andrea-pasquale** [comment]
> Tests on GPU are passing locally both for qibo on qibojit.

Actually after double checking I saw that there were some tests that were failing on GPU in Qibo. The errors are raised also using main. I've fix them in https://github.com/qiboteam/qibo/pull/950/commits/d1559aa5071ff27c61e413019addae24f518808e

**Comment by renatomello** [comment]
> > Tests on GPU are passing locally both for qibo on qibojit.
> 
> Actually after double checking I saw that there were some tests that were failing on GPU in Qibo. The errors are raised also using main. I've fix them in [qiboteam/qibo@d1559aa](https://github.com/qiboteam/qibo/commit/d1559aa5071ff27c61e413019addae24f518808e)

The tests in the quantum_info module were already fixed in https://github.com/qiboteam/qibo/pull/872 and are just waiting on a second reviewer to be merged.

File: pyproject.toml
Comment (stavros11) [inline pyproject.toml:17]: This can now be reverted to main or the latest release since qiboteam/qibo#950 is merged.

**Review by stavros11 — APPROVED** [review]
Thanks for the updates @andrea-pasquale. This should be ready to merged, right?

File: pyproject.toml
Comment (andrea-pasquale) [reply pyproject.toml:17]: Yes, I know. I think that @scarrazza would like to make a new release of qibo soon.
As soon as we release qibo 0.1.15 I will update the lock file.

Comment (scarrazza) [reply pyproject.toml:17]: @andrea-pasquale the new release is now online, could please regenerate the lock file (and test)? Thanks.

Comment (andrea-pasquale) [reply pyproject.toml]: It seems that there are some issues. I will fix them.

Comment (andrea-pasquale) [reply pyproject.toml]: Tests are now passing locally on GPU for both qibo and qibojit (and they seems to pass also on GitHub). We should be ready to merge @scarrazza

Comment (scarrazza) [reply pyproject.toml]: Thanks, feel free to merge when the CI completes. I will then make the new release.

**Pull request merged**
